Speed_Controller/src/Encoder/encoder.c

```diff
@@ -1,68 +1,60 @@
 #include "encoder.h"
-#include "virtual_com.h"
 
 /**
  * Initializes the encoder for the accelerator pedal
  */
-void EncoderInit (void) {
-	
-	//Initialize AFIO Clock
-	RCC->APB2ENR |= 0x1;
-	
-	//TIM 1 Enable
-	RCC->APB2ENR |= 0x1 << 11;
-	
-	//Initialize GPIOA Clock
-	RCC->APB2ENR |= 0x1 << 2;	
-	GPIOA->CRH &= 0;
-	GPIOA->CRH |= 0x00000044;
-	
-	//TIM 1 No Remap
-	AFIO->MAPR &= ~(0x11 << 6);
-	
-	TIM1->CCMR1 |= 0x01;	//Map Channel 1 to TI1
-	TIM1->CCMR1 |= 0x01 << 8; //Map Channel 2 to TI2
-	
-	TIM1->CCER &= ~(0x1 << 1);
-	TIM1->CCER &= ~(0x1 << 3); 
-	
-	TIM1->CCER &= ~(0x1 << 5);
-	TIM1->CCER &= ~(0x1 << 7);
-	
-	TIM1->CCER |= 0x1;
-	TIM1->CCER |= 0x1 << 4;
-	
-	TIM1->SMCR |= 0x011;
-	
-	TIM1->CR1 |= 0x1;	
-	
+void EncoderInit(void)
+{
+	// Initialize AFIO Clock
+	RCC->APB2ENR |= RCC_APB2ENR_AFIOEN;
+
+	// TIM 1 Enable
+	RCC->APB2ENR |= RCC_APB2ENR_TIM1EN;
+
+	// Initialize GPIOA Clock
+	RCC->APB2ENR |= RCC_APB2ENR_IOPAEN;
+	// Configure pins A8 and A9 as inputs
+	GPIOA->CRH &= ~(GPIO_CRH_CNF9 | GPIO_CRH_MODE9 | GPIO_CRH_CNF8 | GPIO_CRH_MODE8);
+	GPIOA->CRH |= (0x1 << GPIO_CRH_CNF9_Pos) | (0x0 << GPIO_CRH_MODE9_Pos) | 
+	              (0x1 << GPIO_CRH_CNF8_Pos) | (0x0 << GPIO_CRH_MODE8_Pos);
+
+	// TIM 1 no pin remap
+	AFIO->MAPR &= ~AFIO_MAPR_TIM1_REMAP;
+
+	TIM1->CCMR1 |= 0x1 << TIM_CCMR1_CC1S_Pos;	// Map capture/compare channel 1 to TI1 (Timer Input 1)
+	TIM1->CCMR1 |= 0x1 << TIM_CCMR1_CC2S_Pos; 	// Map capture/compare channel 2 to TI2 (Timer Input 2)
+
+	// Enable capture/compare channels 1 and 2, other channels disabled and all polarities default
+	TIM1->CCER = (TIM_CCER_CC2E | TIM_CCER_CC1E);
+
+	TIM1->SMCR |= 0x3 << TIM_SMCR_SMS_Pos; // Set timer to encoder mode 3 (clocked by both TI1FP1 and TI2FP2)

```

Comment:
There was an easy and rather significant mistake here - the STM32F1 reference manual specifies a value of '011' for the SMCR register SMS bitfield for encoder mode 3 (where edges of both encoder signals increment/decrement the timer counter). This is the mode we want. The '011' is binary (ie. = 0x3); however, it was entered in the code as hex 0x011 (= 0b0001_0001) instead.

Not only did this put the timer in encoder mode 1 instead of 3, it also changed the TS bitfield unintentionally. The effect of doing this for TIM1 is to enable the trigger output (TRGO signal) of TIM2 as a trigger source. By default it triggers a reset of the TIM1 counter (which tracks the encoder position). This is not good, because TIM2 is used to provide a countdown for the maximum time interval between speed control CAN messages. However, I think by default the TIM2 TRGO is configured to only assert upon software setting the UG bit in TIM2's register, which the software definitely isn't doing. Needs a bit more investigation.